### ISSUE TITLE
Cannot bind double types to prepared statements when using Oracle and DBCP

### ISSUE THREAD

**Issue by Pazus** [jOOQ/jOOQ#7589]
### Expected behavior and actual behavior:
When DBCP2 is used (in tests) to create Datasource, it produces `DelegatingPreparedStatement`. Binding variables may lead to NoSuchMethodException.
Stacktrace:
```
org.jooq.exception.DataAccessException: SQL [null]; Error while writing value at JDBC bind index: 1

	at org.jooq_3.11.0.ORACLE11G.debug(Unknown Source)
	at org.jooq.impl.Tools.translate(Tools.java:2380)
	at org.jooq.impl.AbstractBindContext.bindValue(AbstractBindContext.java:127)
	at org.jooq.impl.Val.accept(Val.java:93)
	at org.jooq.impl.AbstractBindContext.bindInternal(AbstractBindContext.java:254)
	at org.jooq.impl.AbstractBindContext.visit0(AbstractBindContext.java:88)
	at org.jooq.impl.AbstractContext.visit0(AbstractContext.java:446)
	at org.jooq.impl.AbstractContext.visit(AbstractContext.java:207)
	at org.jooq.impl.QueryPartList.accept(QueryPartList.java:119)
	at org.jooq.impl.AbstractBindContext.bindInternal(AbstractBindContext.java:254)
	at org.jooq.impl.AbstractBindContext.visit0(AbstractBindContext.java:88)
	at org.jooq.impl.AbstractContext.visit0(AbstractContext.java:446)
	at org.jooq.impl.AbstractContext.visit(AbstractContext.java:207)
	at org.jooq.impl.AbstractQuery.execute(AbstractQuery.java:346)
	at org.jooq.impl.AbstractResultQuery.fetch(AbstractResultQuery.java:323)
	at org.jooq.impl.AbstractResultQuery.fetch(AbstractResultQuery.java:1455)
	at org.jooq.impl.SelectImpl.fetch(SelectImpl.java:3756)
        ...............................
Caused by: java.sql.SQLException: Error while writing value at JDBC bind index: 1
	at org.jooq.impl.DefaultBindContext.bindValue0(DefaultBindContext.java:67)
	at org.jooq.impl.AbstractBindContext.bindValue(AbstractBindContext.java:124)
	... 71 more
Caused by: org.jooq.tools.reflect.ReflectException: java.lang.NoSuchMethodException: No similar method setBinaryDouble with params [class java.lang.Integer, class java.lang.Double] could be found on type class org.apache.commons.dbcp2.DelegatingPreparedStatement.
	at org.jooq.tools.reflect.Reflect.call(Reflect.java:465)
	at org.jooq.impl.DefaultBinding$DefaultDoubleBinding.set0(DefaultBinding.java:2094)
	at org.jooq.impl.DefaultBinding$DefaultDoubleBinding.set0(DefaultBinding.java:2066)
	at org.jooq.impl.DefaultBinding$AbstractBinding.set(DefaultBinding.java:758)
	at org.jooq.impl.DefaultBindContext.bindValue0(DefaultBindContext.java:62)
	... 72 more
Caused by: java.lang.NoSuchMethodException: No similar method setBinaryDouble with params [class java.lang.Integer, class java.lang.Double] could be found on type class org.apache.commons.dbcp2.DelegatingPreparedStatement.
	at org.jooq.tools.reflect.Reflect.similarMethod(Reflect.java:532)
	at org.jooq.tools.reflect.Reflect.call(Reflect.java:462)
	... 76 more
```

### Steps to reproduce the problem (include example code if possible):
When `DefaultBinding` line 2094 is executed on DBCP's `DelegatingPreparedStatement` it fails on reflection as `DelegatingPreparedStatement` doesn't have `setBinaryDouble` methodm which is specific to `OraclePreparedStatement` interface.

DBCP2 version - 2.3.0

### Versions:

- jOOQ: 3.11.0 pro
- Java: 8
- Database (include vendor): Oracle
- OS: win7 x64
- JDBC Driver (include name if inofficial driver): 12.2.0.1

**Comment by lukaseder** [comment]
Thank you very much for reporting. Will look into this ASAP. This week, 3.11.1 is scheduled, so this will definitely make it in there

**Comment by lukaseder** [comment]
I can reproduce this both with DBCP2 2.4.0 and DBCP 1.4. Related issues are: #3957, #5134

**Comment by lukaseder** [comment]
Fixed for jOOQ 3.12. Backport scheduled for jOOQ 3.11.1: #7592. 3.10 is not affected by this issue.

**Comment by lukaseder** [comment]
Note, this is a regression introduced by #7548
